[MV-426] Porting & Benchmark Sigmoid Focal Loss kernel

BuiChiTrung opened the pull request:
- [x] Port 4 kernels to MiOpen: Fwd, Bwd, UnreducedFwd, UnreducedBwd.
- [x] Add unit tests for 4 kernels.
- [x] Add driver tests for 4 kernels.
- [x] Bench mark performance with ROCm: [link](https://morehio-my.sharepoint.com/:x:/r/personal/kyeonghwan_ryu_moreh_io/_layouts/15/Doc.aspx?sourcedoc=%7B38FB5B9B-387C-4AD1-A594-8CC782CB3E90%7D&file=SigmoidFocalLoss.xlsx&action=default&mobileredirect=true)
- [x] Update docxygen documentation.

Reduce : 
|type    |Forward| Backward |
|--------|-------|-------|
|float32 | 2.61  | 4.27 |
|float16 | 2.18 | 3.82 | 
|bfloat16| 2.44  | 4.15 |

Unreduced: 
|type    | Forward | Backward |
|--------|-------|-------------|
|float32 | 5.09  | 3.39 |
|float16 | 4.74  | 3.03 |
|bfloat16| 5.27  | 3.40 |

BuiChiTrung commented:
This PR has several changes compared to my last closed PR #20 for HingeEmbeddingLoss operation and other Loss kernel PRs. In previous PRs, we use 2 problem description structs and 2 invoke params structs for fwdReduced, fwdUnreduced kernels. However in this PR, I merged to 1 problem desc struct and 1 invoke param struct.

BuiChiTrung commented:
Do we have to support computing the gradient for the target tensor in bwd kernels? I see some loss kernels in moDNN don't support whereas some support (BCELoss, ConsineEmbeddingLoss for instance).

et16kr commented:
> Do we have to support computing the gradient for the target tensor in bwd kernels? I see some loss kernels in moDNN don't support whereas some support (BCELoss, ConsineEmbeddingLoss for instance).

I was delayed because I was looking through the history to investigate an issue I wasn't aware of.
When calculating target_grad in PyTorch's loss, we also added target_grad calculation to the modnn loss.
If a loss in PyTorch calculates target_grad, then it is necessary to include it in the modnn loss. 
Conversely, if a loss in PyTorch does not calculate target_grad, then it is not needed in modnn.
https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/FunctionsManual.cpp

It seems reasonable to follow the implementation in modnn. 
If you are unsure whether target_grad calculation is needed but missing, please check the corresponding PyTorch loss function.

et16kr commented on driver/driver.hpp:
```diff
@@ -151,7 +151,8 @@ inline void PadBufferSize(size_t& sz, int datatype_sz)
            "tensorop[fp16], reduce[fp16|fp64], layernorm[bfp16|fp16], sum[bfp16|fp16], "
-           "argmax[bfp16|fp16], groupnorm[bfp16|fp16], cat[bfp16|fp16]\n");
+           "argmax[bfp16|fp16], groupnorm[bfp16|fp16], cat[bfp16|fp16], "
+           "sigmoidfocalloss[bfp16|fp16|fp32] \n");
```
fp32 seems to be a typo.
```bin/MIOpenDriver sigmoidfocallossfp32``` is not work.

BuiChiTrung replied:
Yes, it is a typo. Driver with float32 is invoked like this: `bin/MIOpenDriver sigmoidfocalloss`

BuiChiTrung commented:
> > Do we have to support computing the gradient for the target tensor in bwd kernels? I see some loss kernels in moDNN don't support whereas some support (BCELoss, ConsineEmbeddingLoss for instance).
> 
> I was delayed because I was looking through the history to investigate an issue I wasn't aware of. When calculating target_grad in PyTorch's loss, we also added target_grad calculation to the modnn loss. If a loss in PyTorch calculates target_grad, then it is necessary to include it in the modnn loss. Conversely, if a loss in PyTorch does not calculate target_grad, then it is not needed in modnn. https://github.com/pytorch/pytorch/blob/main/torch/csrc/autograd/FunctionsManual.cpp
> 
> It seems reasonable to follow the implementation in modnn. If you are unsure whether target_grad calculation is needed but missing, please check the corresponding PyTorch loss function.

I figured out that only loss kernel with int/long target tensor, the dtarget is not computed. I have added the code to compute target tensor gradient.

et16kr commented on include/miopen/miopen.h:
```diff
@@ -512,7 +512,7 @@ typedef enum
     miopenActivationPOWER = 6, /*!< Scaled and shifted power \f$(\alpha + \beta * x)^{gamma}\f$ */
     miopenActivationCLIPPEDRELU =
-        7, /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
+        7,                     /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
```
please check clang-format

BuiChiTrung replied:
Fixed it.

et16kr commented on test/gtest/sigmoid_focal_loss.cpp:
```diff
@@ -0,0 +1,301 @@
+
+using namespace sigmoidfocalloss;
+
+#ifdef TEST_FWD_REDUCED
```
There is define TEST_FWD_REDUCED right above it. So it seems meaningless.
Please explain why you did this

BuiChiTrung replied:
While benchmarking, testing, and debugging we may want to run only a specific set of tests instead of all tests. I added these macros to make it easier to enable/disable all tests of a kernel.

et16kr commented on test/gtest/sigmoid_focal_loss.cpp:
```diff
@@ -0,0 +1,301 @@
+#ifdef TEST_FWD_REDUCED
+TEST_P(SigmoidFocalLossForwardTestFloat32, SigmoidFocalLossForwardTest)
+{
+    if(miopen::IsEnabled(ENV(MIOPEN_TEST_ALL)))
```
We have recently received a request from AMD.
Please refer to the following
https://github.com/ROCm/MIOpen/issues/1843
https://github.com/ROCm/MIOpen/pull/2883#discussion_r1616401109

BuiChiTrung replied:
I have updated the test invoke condition

et16kr commented:
Have you measured performance after adding Target grad?
I think it would be better to optionally select target grad in driver.(default: off)
This is because grad is not generally used in target tensor,
and performance measurements may vary if target grad is included. 
(I think gtest is fine as is.)

I think you made the right choice to include target grad.
But I'll ask others as well.(Their response is OK!!)

et16kr commented on src/include/miopen/tensor_view_utils.hpp:
```diff
@@ -0,0 +1,80 @@
+}
+
+template <int N>
+inline void slice_tv(tensor_view_t<N>& tensor_view, int32_t sliceCount, const int32_t* slices)
```
slice_tv is not used in this PR.
If it was used in another PR, could you please let me know so I can refer to it?

BuiChiTrung replied:
It's used in Seungman's [PR](https://github.com/ROCm/MIOpen/pull/2883/files#diff-ccc972090dd3e61487d0ad364fb34d4aaf8f7c6911c878988654fe7173f3058f)

et16kr replied:
OK I see it.

et16kr commented on driver/sigmoid_focal_loss_driver.hpp:
```diff
@@ -0,0 +1,772 @@
+int SigmoidFocalLossDriver<TIO>::VerifyBackward()
+{
+    RunBackwardCPU();
+#ifdef DEBUGGING
```
Please delete the debugging output.
Please let me know if you still need it, and please delete it when uploading it to AMD.

BuiChiTrung replied:
It's is removed

et16kr commented on src/kernels/MIOpenSigmoidFocalLoss.cpp:
```diff
@@ -0,0 +1,329 @@
+}
+
+extern "C" __global__ void SigmoidFocalLossFwd(const IN_OUT_TYPE* input,
+                                               IN_OUT_TYPE* target,
```
Tensors are of the same dtype.
A dtype check of the tensor is required.
Please let me know if I can't find it. (I only found checkSameLength.)

BuiChiTrung replied:
It's my bad. I have added the check code in problem description constructor

BuiChiTrung commented:
> Have you measured performance after adding Target grad? I think it would be better to optionally select target grad in driver.(default: off) This is because grad is not generally used in target tensor, and performance measurements may vary if target grad is included. (I think gtest is fine as is.)
> 
> I think you made the right choice to include target grad. But I'll ask others as well.

The benchmark result for target grad is updated in the PR desc. Also, i have added a flag to disable/enable target grad in driver, this required changing the tensor_view_utils.hpp as follow:
```
for(size_t i = 0; i < N; ++i)
    {
        if(dims.empty())
        {
            tensor_view.stride[i] = 0;
            tensor_view.size[i]   = 0;
        }
        ...
    }
```

et16kr commented on driver/sigmoid_focal_loss_driver.hpp:
```diff
@@ -0,0 +1,774 @@
+    gamma        = inflags.GetValueDouble("gamma");
+    isContiguous = inflags.GetValueInt("is-contiguous") == 1 ? true : false;
+    isTargetGradientComputed = inflags.GetValueInt("target-gradient") == 1 ? true : false;
+    reduction    = static_cast<miopenLossReductionMode_t>(inflags.GetValueInt("reduction"));
```
clang-format is required.
If you use a github hook, it will be checked every time you commit.
```
./.githooks/install
git config core.hooksPath .githooks
```

BuiChiTrung replied:
I migrated code to a new node yesterday and forgot to setup the githooks and clang-format in IDE. I have updated it.

et16kr commented on test/gtest/sigmoid_focal_loss.hpp:
```diff
@@ -0,0 +1,495 @@
+        SigmoidFocalLossTestCase({100, 500}),        // 2D cont
+        SigmoidFocalLossTestCase({100, 500}, false), // 2D non-cont
+        // SigmoidFocalLossTestCase({2, 2}), // 2D non-cont
+        // SigmoidFocalLossTestCase({2, 2}, false), // 2D non-cont
```
If you do not use it, please remove it.

BuiChiTrung replied:
It is removed.

et16kr commented on include/miopen/miopen.h:
```diff
@@ -4950,6 +4950,98 @@ MIOPEN_EXPORT miopenStatus_t miopenCTCLoss(miopenHandle_t handle,
+                                              miopenLossReductionMode_t reduction,
+                                              size_t* sizeInBytes);
+
+/*! @brief Execute a SigmoidFocalLoss reduced forward layer
```
Since this is an API that includes both reduced and unreduced, 
I think 'reduced' should be removed from the function comment.

et16kr commented on include/miopen/miopen.h:
```diff
@@ -4950,6 +4950,98 @@ MIOPEN_EXPORT miopenStatus_t miopenCTCLoss(miopenHandle_t handle,
+                                              miopenLossReductionMode_t reduction,
+                                              size_t* sizeInBytes);
+
+/*! @brief Execute a SigmoidFocalLoss reduced forward layer
```
This is an API that includes both reduced and unreduced.
I think "reduced" should be removed in function comment.

BuiChiTrung replied:
It is removed.

et16kr commented on test/gtest/sigmoid_focal_loss.hpp:
```diff
@@ -0,0 +1,495 @@
+        std::fill(dInputHost.begin(), dInputHost.end(), 0);
+
+        dTarget = tensor<TIO>{in_dims};
+        std::fill(dInput.begin(), dInput.end(), 0);
```
maybe ```std::fill(dTarget.begin(), dTarget.end(), 0);```

BuiChiTrung replied:
You're right. Updated

et16kr commented on test/gtest/sigmoid_focal_loss.hpp:
```diff
@@ -0,0 +1,495 @@
+        std::fill(dInputHost.begin(), dInputHost.end(), 0);
+
+        dTarget = tensor<TIO>{in_dims};
+        std::fill(dInput.begin(), dInput.end(), 0);
```
```std::fill(dTarget.begin(), dTarget.end(), 0);```

BuiChiTrung replied:
You're right. Updated

BuiChiTrung commented on include/miopen/miopen.h:
```diff
@@ -512,7 +512,7 @@ typedef enum
     miopenActivationPOWER = 6, /*!< Scaled and shifted power \f$(\alpha + \beta * x)^{gamma}\f$ */
     miopenActivationCLIPPEDRELU =
-        7, /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
+        7,                     /*!< Clipped Rectified Linear Unit \f$ min(\alpha, max(0,x)) \f$ */
```
My IDE keeps adding this empty space, as my VPN suddenly stopped working, I will fix it tomorrow.

et16kr commented:
Thank you for your hard work.
